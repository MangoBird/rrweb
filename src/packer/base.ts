import { eventWithTime } from '../types';

export abstract class Packer {
  public version!: number;

  public pack!: (events: eventWithTime[]) => string;
  public unpack!: (raw: string) => eventWithTime[];
}

export type PackedData<T> = {
  meta: Meta;
  data: T;
};

export type Meta = {
  packer: string;
  version: number;
};
