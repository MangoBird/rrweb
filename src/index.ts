import record from './record';
import { Replayer } from './replay';
import { mirror } from './utils';
import { pack, unpack } from './packer';

export {
  EventType,
  IncrementalSource,
  MouseInteractions,
  ReplayerEvents,
} from './types';

const { addCustomEvent } = record;

export { record, addCustomEvent, Replayer, mirror };
