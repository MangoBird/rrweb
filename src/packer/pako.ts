/* tslint:disable: no-console */
import { deflate, inflate } from 'pako';
import { eventWithTime } from '../types';
import { Packer, PackedData } from './base';

export class PakoPacker implements Packer {
  public version = 1;

  public pack(events: eventWithTime[]): string {
    const packedString = deflate(JSON.stringify(events), { to: 'string' });
    const data: PackedData<string> = {
      meta: {
        packer: 'pako',
        version: this.version,
      },
      data: packedString,
    };
    return JSON.stringify(data);
  }

  public unpack(raw: string): eventWithTime[] {
    const data: PackedData<string> = JSON.parse(raw);
    if (!data.meta && Array.isArray(data)) {
      // unpacking unpacked data, maybe legacy events
      console.info('unpacking unpacked data...');
      return data;
    }
    if (!data.meta) {
      throw new Error('Unknown data format.');
    }
    if (data.meta.packer !== 'pako') {
      console.info(data.meta);
      throw new Error('These events were not packed by the pako packer.');
    }
    if (data.meta.version !== this.version) {
      throw new Error(
        `These events were packed with version ${data.meta.version} which is incompatible with current version ${this.version}.`,
      );
    }
    return JSON.parse(inflate(data.data, { to: 'string' }));
  }
}
