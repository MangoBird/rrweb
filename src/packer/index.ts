import { PakoPacker } from './pako';

const pakoPackger = new PakoPacker();

const { pack, unpack } = pakoPackger;

export { pack, unpack };
