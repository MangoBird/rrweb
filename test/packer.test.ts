import { expect } from 'chai';
import { matchSnapshot } from './utils';
import { pack, unpack } from '../src/packer';
import { eventWithTime, EventType } from '../src/types';

const events: eventWithTime[] = [
  {
    type: EventType.DomContentLoaded,
    data: {},
    timestamp: new Date('2020-01-01').getTime(),
  },
];

describe('pack', () => {
  it('can pack events', () => {
    const packedData = pack(events);
    matchSnapshot(packedData, __filename, 'pack');
  });
});

describe('unpack', () => {
  it('is compatible with unpacked data 1', () => {
    const result = unpack((events as unknown) as string);
    expect(result).to.deep.equal(events);
  });

  it('is compatible with unpacked data 2', () => {
    const result = unpack(JSON.stringify(events));
    expect(result).to.deep.equal(events);
  });

  it('stop on unknown data format', () => {
    expect(() => unpack('{}')).to.throw('Unknown data format.');
  });

  it('stop on unmatched packer', () => {
    expect(() =>
      unpack(
        JSON.stringify({
          meta: {
            packer: 'dummy',
          },
        }),
      ),
    ).to.throw('These events were not packed by the pako packer.');
  });

  it('stop on unmatched packer version', () => {
    expect(() =>
      unpack(
        JSON.stringify({
          meta: {
            packer: 'pako',
            version: 2,
          },
        }),
      ),
    ).to.throw(/incompatible with current version/);
  });

  it('can unpack packed data', () => {
    const packedData = pack(events);
    const result = unpack(packedData);
    expect(result).to.deep.equal(events);
  });
});
