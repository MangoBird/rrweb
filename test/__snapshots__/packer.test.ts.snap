// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pack 1`] = `"{\\"meta\\":{\\"packer\\":\\"pako\\"},\\"data\\":\\"x®V*©,HU²2ÐQJI,IT²ª®ÕQ*ÉÌM-.IÌ-P²2457·06³0\\\\u0000ÚX\\\\u0000v¯\\\\u000ex\\"}"`;
